File: src-tauri/src/commands/roles.rs
//! Tauri commands backing `~/.cortex/roles/*.yaml` — pre-built agent personas.
//!
//! Personas (a.k.a. "roles") are pure data. The only side-effect command in
//! here is `apply_role_to_agent`, which translates a role's `system_prompt`
//! into the existing per-agent custom-instructions mechanism so the rest of
//! the chat pipeline picks it up automatically.

use crate::agents::roles::{self, Role};
use crate::orchestrator;

/// List every role under `~/.cortex/roles/*.yaml`, sorted by name. Missing
/// directory yields `[]` — never an error.
#[tauri::command]
pub async fn list_roles() -> Result<Vec<Role>, String> {
    Ok(roles::list_roles())
}

/// Load a single role by filename stem.
#[tauri::command]
pub async fn get_role(name: String) -> Result<Role, String> {
    if name.trim().is_empty() {
        return Err("name is required".into());
    }
    roles::get_role(&name).ok_or_else(|| format!("role '{name}' not found"))
}

/// Create or update a role on disk. Returns the role as persisted.
#[tauri::command]
pub async fn set_role(role: Role) -> Result<Role, String> {
    roles::set_role(&role).map_err(|e| e.to_string())?;
    Ok(role)
}

/// Delete a role file. Missing files are a no-op.
#[tauri::command]
pub async fn delete_role(name: String) -> Result<(), String> {
    if name.trim().is_empty() {
        return Err("name is required".into());
    }
    roles::delete_role(&name).map_err(|e| e.to_string())
}

/// Apply a role's `system_prompt` to `agent_id` by writing it into the
/// per-agent custom instructions store. The chat pipeline already composes
/// those into the outgoing system prompt — no other plumbing required.
///
/// Returns the trimmed prompt that landed on disk so the UI can confirm.
#[tauri::command]
pub async fn apply_role_to_agent(
    role_name: String,
    agent_id: String,
) -> Result<String, String> {
    if role_name.trim().is_empty() {
        return Err("role_name is required".into());
    }
    if agent_id.trim().is_empty() {
        return Err("agent_id is required".into());
    }
    let role = roles::get_role(&role_name)
        .ok_or_else(|| format!("role '{role_name}' not found"))?;
    let prompt = role.system_prompt.unwrap_or_default();
    orchestrator::set_agent_instructions(&agent_id, &prompt).map_err(|e| e.to_string())
}

File: src/lib/roles.ts
import { invoke } from "@tauri-apps/api/core";

/**
 * Mirror of `agents::roles::Role`. A "role" is a re-usable agent persona
 * stored at `~/.cortex/roles/<name>.yaml`. Optional fields are omitted from
 * the wire format when `None` on the Rust side.
 */
export interface Role {
  name: string;
  description?: string;
  tools?: string[];
  model?: string;
  system_prompt?: string;
}

/** List every role under `~/.cortex/roles/*.yaml`. Returns `[]` when empty. */
export async function listRoles(): Promise<Role[]> {
  return invoke<Role[]>("list_roles");
}

/** Load a single role by filename stem. Throws on missing / malformed. */
export async function getRole(name: string): Promise<Role> {
  return invoke<Role>("get_role", { name });
}

/** Create or update a role on disk. Returns the persisted role. */
export async function setRole(role: Role): Promise<Role> {
  return invoke<Role>("set_role", { role });
}

/** Delete a role file. Missing files are a no-op. */
export async function deleteRole(name: string): Promise<void> {
  return invoke("delete_role", { name });
}

/**
 * Apply a role's `system_prompt` to the agent identified by `agentId`. Routes
 * through the existing per-agent custom-instructions storage so the chat
 * pipeline picks it up automatically. Returns the prompt as persisted (empty
 * string if the role had no system prompt).
 */
export async function applyRoleToAgent(
  roleName: string,
  agentId: string,
): Promise<string> {
  return invoke<string>("apply_role_to_agent", { roleName, agentId });
}

// ── Profile bundling (Codex #10) ────────────────────────────────────────────

/**
 * Per-dimension report from `apply_profile_v2`. Each entry in `applied` names
 * a dimension that was successfully bundled (model / sandbox_mode /
 * approval_policy / reasoning_effort). `errors` collects per-dimension
 * failure messages so partial success is visible to the UI.
 */
export interface ProfileApplyResult {
  applied: string[];
  errors: string[];
  name: string;
}

/**
 * Bundle-apply a profile: model + sandbox mode + approval policy + reasoning
 * effort in a single call. The classic `apply_profile` only mutates AppState;
 * the v2 variant additionally persists sandbox + approval policy to disk so
 * they survive a relaunch.
 */
export async function applyProfileV2(
  projectRoot: string,
  name: string,
): Promise<ProfileApplyResult> {
  return invoke<ProfileApplyResult>("apply_profile_v2", { projectRoot, name });
}
